libsignal-service-dotnet/SignalServiceMessagePipe.cs

```diff
@@ -99,18 +103,17 @@ internal SignalServiceMessagePipe(CancellationToken token, SignalWebSocketConnec
                 Body = ByteString.CopyFrom(Encoding.UTF8.GetBytes(JsonUtil.ToJson(list)))
             };
             requestmessage.Headers.Add("content-type:application/json");
-            Logger.LogDebug("Sending message {0}", requestmessage.Id);
-            var t = Websocket.SendRequest(requestmessage);
-            await t;
-            if (t.IsCompleted)
+            var sendTask = Websocket.SendRequest(requestmessage).Task;
+            if (await Task.WhenAny(sendTask, Task.Delay(10*1000)) == sendTask)

```

Comment:
You should add a cancellation token to the Task.Delay so that it gets cancelled if the request is successful and we don't have a bunch of timers laying around: https://stackoverflow.com/a/22078975/6681022